### PR TITLE
xray client: return an error if the HTTP request failed

### PR DESCRIPTION
**Pull request by jaedle**

### CODE REVIEW

File: samplers/aws/xray/internal/client.go [hunk 1]
```diff
@@ -139,6 +139,10 @@ func (c *xrayClient) getSamplingRules(ctx context.Context) (*getSamplingRulesOut
 	}
 	defer output.Body.Close()
 
+	if output.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("xray client:  unable to retrieve sampling settings, expected response status code 200, got: %d", output.StatusCode)
```
Comment (dmathieu) [inline samplers/aws/xray/internal/client.go:143]:
```suggestion
		return nil, fmt.Errorf("xray client: unable to retrieve sampling settings, expected response status code 200, got: %d", output.StatusCode)
```

File: samplers/aws/xray/internal/client.go [hunk 2]
```diff
@@ -139,6 +139,10 @@ func (c *xrayClient) getSamplingRules(ctx context.Context) (*getSamplingRulesOut
 	}
 	defer output.Body.Close()
 
+	if output.StatusCode != http.StatusOK {
```
Comment (dmathieu) [inline samplers/aws/xray/internal/client.go:142]: Should this accept any 2xx HTTP responses?

**Review by dmathieu — COMMENTED** [review]
This needs a changelog entry

File: samplers/aws/xray/internal/client.go (hunk 2 above)
Comment (jaedle) [reply samplers/aws/xray/internal/client.go:142]: TBH I asked myself the same question. As the docs only state a successful response should be 200 OK, I would argue that checking for that exact response code should be perfectly fine.
If you think that should be change, feel free to response, so I will change this :)

File: samplers/aws/xray/internal/client.go (hunk 1 above)
Comment (jaedle) [reply samplers/aws/xray/internal/client.go]: done

File: CHANGELOG.md [hunk 3]
```diff
@@ -16,6 +16,7 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
 
 - The superfluous `response.WriteHeader` call in `go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp` when the response writer is flushed. (#5634)
 - Custom attributes targeting metrics recorded by the `go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp` are not ignored anymore. (#5129)
+- Erroneous response code of `getSamplingRules` in `go.opentelemetry.io/contrib/samplers/aws/xray/internal/client.go` will result in an error despite an empty sampling manifest. (#)
```
Comment (dmathieu) [inline CHANGELOG.md:19]:
```suggestion
- Non-200 HTTP status codes when retrieving sampling rules in `go.opentelemetry.io/contrib/samplers/aws/xray/internal/client.go` now return an error. (#5718)
```

**Comment by dmathieu** [comment]
cc @wangzlei as the soon to be owner of this instrumentation.

File: samplers/aws/xray/internal/client_test.go [hunk 4]
```diff
@@ -261,3 +261,22 @@ func TestEndpointIsNotReachable(t *testing.T) {
...
+	t.Cleanup(testServer.Close)
+
+	u, err := url.Parse(testServer.URL)
+	require.NoError(t, err)
+
+	client, err := newClient(*u)
+	require.NoError(t, err)
+
+	samplingRules, err := client.getSamplingRules(context.Background())
+	require.Error(t, err)
```
Comment (dmathieu) [inline samplers/aws/xray/internal/client_test.go:280]: Could you add an assertions checking that the error content matches the proper error? So if somehow, another error is returned, this doesn't still pass.

Comment (jaedle) [reply samplers/aws/xray/internal/client_test.go:280]: Sure, no problem.
I tried to adapt to the existing tests where errors are not explicitly asserted. 

I would recommend to add more specific assertions on all tests related to errors or even add new ones to distinguish error scenarios. 

If that’s fine for you I would work on that.

**Comment by jaedle** [comment]
Thanks for the feedback.

Following you gave a thumbs up on my comment above, I thought that would be appropriate.

> If that’s fine for you, I would work on that within this PR.

Sorry for my confusion.

**Comment by dmathieu** [comment]
My comment was about assertion of errors in tests.
You're also changing the errors content, and changing the way the response body is closed now.

**Comment by jaedle** [comment]
> My comment was about assertion of errors in tests. You're also changing the errors content, and changing the way the response body is closed now.

Rolled back in a moment, thanks.

**Comment by jaedle** [comment]
Sorry for the change, the old way of dealing with response bodies was restored.

**Review by dmathieu — APPROVED** [review]

**Comment by jaedle** [comment]
Hey :wave: 

Any chance to get this merged and released realistically in foreseeable future?

**Comment by dmathieu** [comment]
This needs a second review from a @open-telemetry/go-approvers

**Comment by jaedle** [comment]
Anything I can do from my side to speed things up? It's already 3 month in review.

**Review by dashpole — APPROVED** [review]

File: CHANGELOG.md [hunk 5]
```diff
@@ -102,6 +102,7 @@ The next release will require at least [Go 1.22].
 - The superfluous `response.WriteHeader` call in `go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp` when the response writer is flushed. (#5634)
 - Use `c.FullPath()` method to set `http.route` attribute in `go.opentelemetry.io/contrib/instrumentation/github.com/gin-gonic/gin/otelgin`. (#5734)
 - Out-of-bounds panic in case of invalid span ID in `go.opentelemetry.io/contrib/propagators/b3`. (#5754)
+- Non-200 HTTP status codes when retrieving sampling rules in `go.opentelemetry.io/contrib/samplers/aws/xray` now return an error. (#5718)
```
Comment (dashpole) [inline CHANGELOG.md:105]: this needs to be moved up to the unreleased section now

**Comment by dashpole** [comment]
Sorry.  We don't have any codeowners for the component, so reviews and such are likely to be much slower.

**Pull request merged**